wolfcrypt/src/asn.c

```diff
@@ -13743,190 +13746,99 @@ static int MatchBaseName(int type, const char* name, int nameSz,
-                    /* allow permitted dirName smaller than actual subject */
-                    if (cert->subjectRawLen >= base->nameSz &&
-                        XMEMCMP(cert->subjectRaw, base->name,
-                                                        base->nameSz) == 0) {
-                        return 0;
-                    }
-                    #ifndef WOLFSSL_NO_ASN_STRICT
-                    /* RFC 5280 section 4.2.1.10
-                       "Restrictions of the form directoryName MUST be
-                        applied to the subject field .... and to any names
-                        of type directoryName in the subjectAltName
-                        extension"
-                    */
-                    if (cert->altDirNames != NULL) {
-                        DNS_entry* cur = cert->altDirNames;
-                        while (cur != NULL) {
-                            if (XMEMCMP(cur->name, base->name, base->nameSz)
-                                    == 0) {
-                                WOLFSSL_MSG("DIR alt name constraint err");
-                                return 0;
-                            }
-                            cur = cur->next;
-                        }
-                    }
-                    #endif /* !WOLFSSL_NO_ASN_STRICT */
-                    break;
-                }
-                default:
-                    break;
-            }; /* switch */
-            base = base->next;
-        }
-    }
+    if (signer->excludedNames == NULL && signer->permittedNames == NULL)
+        return 1;
 
-    /* Check against the permitted list */
-    if (signer->permittedNames != NULL) {
-        int permittedDir = 0;
-        int matchDir;
+    for (i=0; i < sizeof(nameTypes); i++) {
+        byte nameType = nameTypes[i];
+        DNS_entry* name = NULL;
+        DNS_entry  subjectDnsName;
         Base_entry* base;
-        DNS_entry* name;
 
-        /* Check each DNS name matches a permitted. */
-        name = cert->altNames;
-        while (name != NULL) {
-            int matchDns = 0;
-            int permittedDns = 0;
-            base = signer->permittedNames;
-            do {
-                /* Looking for permittedNames that are for DNS. */
-                if (base->type == ASN_DNS_TYPE) {
-                    permittedDns = 1;
-                    matchDns = MatchBaseName(ASN_DNS_TYPE,
-                                             name->name, name->len,
-                                             base->name, base->nameSz);
-                }
-                base = base->next;
-            } while (base != NULL && !matchDns);
-            /* If we found an DNS type permittedName then name must have had a
-             * match. */
-            if (permittedDns && !matchDns)
-                return 0;
-
-            if (!permittedDns)
+        switch (nameType) {
+            case ASN_DNS_TYPE:
+                /* Should it also consider CN in subject? It could use
```

Comment:
Existing code only checks altNames. Are you able to find examples where subjectCN or subjectDnsName could be used per spec?

Reply:
"   If a subjectAltName extension of type dNSName is present, that MUST
   be used as the identity. Otherwise, the (most specific) Common Name
   field in the Subject field of the certificate MUST be used. Although
   the use of the Common Name is existing practice, it is deprecated and
   Certification Authorities are encouraged to use the dNSName instead."
https://datatracker.ietf.org/doc/html/rfc2818

"   Subject alternative names MAY be constrained in the same manner as
   subject distinguished names using the name constraints extension as
   described in Section 4.2.1.10."

It is the opposite of what wolfssf does. The default would be to validate subject DN and also use the same validation to altNames.

However, WolfSSL and other client/implementations (like chrome https://www.pei.com/chrome-58-common-name/) started to ignore the CN field. OpenSSL, for example, checks CN (if it is a valid DNS entry) and e-mail from subject as well as all subjaltnames.
https://github.com/openssl/openssl/blob/485d0790ac1a29a0d4e7391d804810d485890376/crypto/x509/v3_ncons.c#L312

If it is a lib user decision to use or not a certificate CN as dns entry (as well as E for email), wolfssl must validate it beforehand. If it is a exclusive decision for the SSL lib, it might be OK to skip subject DN fields from name constraint (although it violates the spec).

In my code, it would be easy to add those checks. Just prepend the name list with a static entry for CN or E if present in subject DN just like I did with altDirName. The complex code would be to validate if CN is a valid DNS entry (because it should not validate a CN that cannot be used as a DNS entry).